plugin-server/src/utils/lazy-loader.ts

```diff
@@ -0,0 +1,150 @@
+    private lastUsed: Record<string, number | undefined>
+    private cacheUntil: Record<string, number | undefined>
+
+    constructor(private readonly options: LazyLoaderOptions<T>) {
+        this.cache = {}
+        this.lastUsed = {}
+        this.cacheUntil = {}
+    }
+
+    /**
+     * Ensure that a range of values are preloaded and cached.
+     *
+     * If already cached, the lastUsed value is updated to now
+     * If not cached, the value is loaded as part of the batch and added to the cache.
+     * If the value is older than the refreshAge, it is loaded from the database.
+     */
+    public async load(keys: string[]): Promise<Record<string, T | null>> {
+        const results: Record<string, T | null> = {}
+
+        const {
+            loader,
+            refreshAge = REFRESH_AGE,
+            refreshNullAge = REFRESH_AGE,
+            refreshJitterMs = REFRESH_JITTER_MS,
+            throwOnLoadError = true,
+        } = this.options
+        const keysToLoad = new Set<string>()
+
+        // First, check if all keys are already cached and update the lastUsed time
+        for (const key of keys) {
+            const cached = this.cache[key]
+
+            if (cached !== undefined) {
+                results[key] = cached
+                // Always update the lastUsed time
+                this.lastUsed[key] = Date.now()
+
+                const cacheUntil = this.cacheUntil[key] ?? 0
+
+                if (Date.now() > cacheUntil) {
+                    keysToLoad.add(key)
+                    lazyLoaderCacheHits.labels({ name: this.options.name, hit: 'miss' }).inc()
+                    continue
+                }
+            } else {
+                keysToLoad.add(key)
+                lazyLoaderCacheHits.labels({ name: this.options.name, hit: 'miss' }).inc()
+                continue
+            }
+
+            lazyLoaderCacheHits.labels({ name: this.options.name, hit: 'hit' }).inc()
+        }
+
+        if (keysToLoad.size === 0) {
+            lazyLoaderFullCacheHits.labels({ name: this.options.name, hit: 'hit' }).inc()
+            return results
+        }
+
+        lazyLoaderFullCacheHits.labels({ name: this.options.name, hit: 'miss' }).inc()
+
+        let loaded: Record<string, T | undefined>
+        try {
+            loaded = await loader(Array.from(keysToLoad))
+        } catch (error) {
+            if (throwOnLoadError) {
+                throw error
+            }
+            status.error(
+                '🍿',
+                `[LazyLoader](${this.options.name}) Error loading values but silently ignoring: ${error}`
+            )
+            loaded = {}
+        }
+
+        for (const key of keysToLoad) {
+            results[key] = this.cache[key] = loaded[key] ?? null
+            this.cacheUntil[key] =
+                Date.now() +
+                (loaded[key] === null ? refreshNullAge : refreshAge) +
+                Math.floor(Math.random() * refreshJitterMs)
```

Comment:
question: do we want to cache keys for failed loads if throw on load error is disabled? Right now if there's an error and we supress it, all keys to load will be set to null  with full refresh period (`loaded[key]` is undefined).

It feels like that behaviour is not what we'd want, even if we didn't want to raise the errors.

Reply:
It's a bit tricky as there is no user of this now but the intention was that you may have a loader where you don't mind if there are errors. If we would retry every time then you might flood your loader destination.

I can maybe remove the throw thing given there are no users of it yet and therefore hard to agree on a pattern.